Unit test and potential fix for #612.

baumatron opened the pull request:
The issue here is that when performing sequential updates or an insert followed by an update on the server, the server barfs on the second update because the version provided by the server after the first operation isn't saved and therefore isn't sent to the server when attempting the second operation. See #612 for more details.

This change adds a unit test for this case and a potential fix. All unit tests passing with this change.

azurecla commented:
Hi __@baumatron__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.azure.com.
    </span>

TTYL, AZPRBOT;

azurecla commented:
__@baumatron__, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, AZPRBOT;

phvannor commented on sdk/iOS/src/MSQueuePushOperation.m:
```diff
@@ -225,6 +225,43 @@ - (void) finishTableOperation:(MSTableOperation *)operation item:(NSDictionary *
                 }
             }
+            else {
+                // Operation was successful, but another operation is pending.
```
I think you need to compare operation.item with item (minus system properties) for this to be ok to do, as we don't want to stomp on other server changes.

baumatron replied:
This would seem to be the place to hook in a local change conflict delegate to resolve any conflicts and allow for preserving server side changes. What do you think about adding something like that here?

phvannor replied:
This is a logical place to handle a conflicts during a push with a saved op.  That doesn't help anything already in memory in the app. 

I'd like to see the logic I outlined as the fallback case I think either way.

If adding a conflict handling delegate method it would ideally be usable from a SyncTable action, and easily called from the tableOperation: onComplete: method as well as here.  Also, potentially from a item change notification during Pulls & Pushes.

phvannor commented on sdk/iOS/src/MSQueuePushOperation.m:
```diff
@@ -225,6 +225,43 @@ - (void) finishTableOperation:(MSTableOperation *)operation item:(NSDictionary *
+                NSError* storeError;
+                
+                // Get the item from the local store, which contains updates that haven't yet been pushed to the server.
+                NSDictionary* storeItem = [self.syncContext.dataSource readTable:operation.tableName
```
In the case where the pending operation is a delete, there will be no storeItem.  What you actually want to update is the item in the operation queue.  

You aren't explicitly handling a storeItem = nil case in the below logic.

baumatron replied:
Makes sense.

phvannor commented on sdk/iOS/src/MSQueuePushOperation.m:
```diff
@@ -225,6 +225,43 @@ - (void) finishTableOperation:(MSTableOperation *)operation item:(NSDictionary *
+                    NSString* newVersion = [item objectForKey:MSSystemColumnVersion];
+                    [storeItem setValue:newVersion forKey:MSSystemColumnVersion];
+                    
+                    [self.syncContext.dataSource upsertItems:[NSArray arrayWithObject:storeItem]
```
Continuing the above, in the pending delete case, this will cause errors.

phvannor commented on sdk/iOS/test/MSSyncTableTests.m:
```diff
@@ -340,6 +403,146 @@ -(void) testInsertPushInsertPush
 
+
+
+-(void) testInsertUpdateWhilePushing
```
Ideally we'd end up with 2 unit tests: update with server changes (leaves version alone), this test case (Server item == sent copy), and a new one for pending delete gets version updated (if server == sent).

phvannor commented:
Hey Brian, can you let me know if you pla to address my comments or not?  If not, I may take some time to look at adding the approach in myself over the next week or two.

baumatron commented:
Sure, I can update the pull request addressing your feedback. Let me know what you think about the local merge conflict handler and I can throw something together for that.

phvannor commented:
I think keeping the local conflict handler as a separate PR may be cleaner so we can take the changes quicker.  I imagine agreeing on conflict handling may be more involved, esp around figuring out how the work will apply to the other SDKs.

paulbatum commented:
@phvannor this PR is pretty old, where did we end up here? Should we close?

phvannor commented:
we will have to close this one, but I wanted some of the core concepts to be carried over, but part of it is tied up in a pull handler and/or hooks for post pull/push to resolve unecessary conflicts down stream.  With Naren's work on hold, I guess we hold for reference, and possibly refactor to take a couple of the smaller ideas in when time permits.

adrianhall commented:
Closing this PR as the Azure Mobile Services repo is frozen in preference of the Azure Mobile Apps work.
